quinn-proto/src/connection/mod.rs

```diff
@@ -2585,15 +2624,25 @@ impl Connection {
                     close = Some(reason);
                 }
                 Frame::PathChallenge(token) => {
-                    if self
-                        .path_response
-                        .as_ref()
-                        .map_or(true, |x| x.packet <= number)
-                    {
-                        self.path_response = Some(PathResponse {
-                            packet: number,
-                            token,
-                        });
+                    /// Arbitrary permissive limit to prevent abuse
+                    const MAX_PATH_RESPONSES: usize = 16;
+                    let response = PathResponse {
+                        packet: number,
+                        token,
+                    };
+                    let existing = self
+                        .path_responses
+                        .iter_mut()
+                        .find(|(_, addr)| *addr == remote);
+                    if let Some((existing, _)) = existing {
+                        // Update an queued response
+                        if existing.packet <= number {
+                            *existing = response;
+                        }
+                    } else if self.path_responses.len() < MAX_PATH_RESPONSES {
+                        self.path_responses.push((response, remote));
+                    } else {
+                        trace!("ignoring excessive PATH_CHALLENGE");
```

Comment:
How do you feel about wrapping this up in a `PathResponses` type? It feels like there is substantial logic on both the push and the pop case which it might be nice to outline. (If we store this in `PathData`, maybe it could be methods there instead of an additional type.)

Reply:
Seems reasonable. Done.